File: python/problems/leetcode/0238_product_of_array_except_self/solution.py
"""LeetCode 238: Product of Array Except Self."""

from __future__ import annotations


class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        n = len(nums)
        out = [1] * n
        for i in range(1, n):
            out[i] = out[i - 1] * nums[i - 1]
        right = 1
        for i in range(n - 1, -1, -1):
            out[i] *= right
            right *= nums[i]
        return out


if __name__ == "__main__":
    assert Solution().productExceptSelf([1, 2, 3, 4]) == [24, 12, 8, 6]
    assert Solution().productExceptSelf([-1, 1, 0, -3, 3]) == [0, 0, 9, 0, 0]
    print("0238 OK")

File: cpp/problems/leetcode/0238_product_of_array_except_self/solution.cpp
// LeetCode 238
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = (int)nums.size();
        vector<int> out(n, 1);
        for (int i = 1; i < n; ++i) out[i] = out[i - 1] * nums[i - 1];
        int right = 1;
        for (int i = n - 1; i >= 0; --i) {
            out[i] *= right;
            right *= nums[i];
        }
        return out;
    }
};

int main() {
    vector<int> a{1, 2, 3, 4};
    assert((Solution().productExceptSelf(a) == vector<int>{24, 12, 8, 6}));
    vector<int> b{-1, 1, 0, -3, 3};
    assert((Solution().productExceptSelf(b) == vector<int>{0, 0, 9, 0, 0}));
    cout << "0238 OK" << endl;
    return 0;
}
